frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java

```diff
@@ -216,67 +213,57 @@ private void doAutomaticDiscover(Path macheteFilePath) {
-      @Override
-      public void onSuccess() {
-        discoverResult.onFailure(exception -> {
-          Class<? extends Throwable> exceptionClass = exception.getClass();
-          if (EmptyRepositoryException.class.equals(exceptionClass)) {
-            setTextForEmptyTable(format(getString("string.GitMachete.EnhancedGraphTable.empty-machete-file.upper-text"),
-                macheteFilePath.toString()));
-          } else if (EmptyBranchLayoutException.class.equals(exceptionClass)) {
-            LOG.error("Can't get branch layout from repository snapshot");
-          } else {
-            VcsNotifier.getInstance(project).notifyError(
+        var discoverRunResult = Try.of(() -> RuntimeBinding.instantiateSoleImplementingClass(IGitMacheteRepositoryCache.class)
+            .getInstance(mainDirPath, gitDirPath).discoverLayoutAndCreateSnapshot());
+
+        if (!discoverRunResult.isSuccess()) {
+          var exception = discoverRunResult.getCause();
+          GuiUtils.invokeLaterIfNeeded(() -> VcsNotifier.getInstance(project).notifyError(
+              getString("string.GitMachete.EnhancedGraphTable.automatic-discover.cant-discover-layout-error-title"),
+              exception.getMessage() != null ? exception.getMessage() : ""), NON_MODAL);
+          return;
+        }
+
+        var repositorySnapshot = discoverRunResult.get();
+
+        if (repositorySnapshot.getRootBranches().size() == 0) {
+          GuiUtils.invokeLaterIfNeeded(
+              () -> setTextForEmptyTable(format(getString("string.GitMachete.EnhancedGraphTable.empty-machete-file.upper-text"),
+                  macheteFilePath.toString())),
+              NON_MODAL);
+          return;
+        }
+
+        var branchLayoutWriter = project.getService(BranchLayoutWriterProvider.class).getBranchLayoutWriter();
+        var branchLayout = repositorySnapshot.getBranchLayout().getOrNull();
+
+        if (branchLayout == null) {
+          LOG.error("Can't get branch layout from repository snapshot");
+          return;
+        }
+
+        @SuppressWarnings("guieffect") // To allow lambda invocation in `onSuccess` method
+        var __ = Try.run(() -> branchLayoutWriter.write(macheteFilePath, branchLayout, /* backupBranchLayout */ true))
+            .onFailure(exception -> GuiUtils.invokeLaterIfNeeded(() -> VcsNotifier.getInstance(project).notifyError(
                 getString("string.GitMachete.EnhancedGraphTable.automatic-discover.cant-discover-layout-error-title"),
-                exception.getMessage() != null ? exception.getMessage() : "");
-          }
-        })
-            .onSuccess(result -> {
-              VcsNotifier.getInstance(project)
-                  .notifyInfo(getString("string.GitMachete.EnhancedGraphTable.automatic-discover.success-message"));
-
-              gitMacheteRepositorySnapshot = result.repositorySnapshot;
+                exception.getMessage() != null ? exception.getMessage() : ""), NON_MODAL))
+            .onSuccess(v -> {
```

Comment:
Nit:
```suggestion
            .onSuccess(__ -> {
```

Reply:
Probably should be even `___` (triple `_`) coz double `_` is already used in this scope

Reply:
Reworked, no longer relevant